### PR TITLE
Refactor deletefolder API

### PR DESCRIPTION
**Pull request by Tulsishah**
### Description

- Added the includeFolderAsPrefix parameter with a value of true, as required for HNS scenarios. We internally pass  `includeFolderAsPrefix` true in ListObject method.
- Refactored deleteFolder API. As previous logic was not logging DeleteObject request in case on FLAT bucket properly.
- Added unit test for prefix_bucket_test
- Also, included the `enable-hns` flag in the implicit-dir package to enable end-to-end testing for the delete folder functionality.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - Automated

### CODE REVIEW

File: internal/storage/bucket_handle.go [hunk 1]
```diff
@@ -483,6 +483,7 @@ func (b *bucketHandle) DeleteFolder(ctx context.Context, folderName string) (err
 		if !errors.As(err, &notfound) {
 			return err
 		}
+		logger.Infof("Object in DeleteObject not found for delete folder: %v", err)
```
Comment (ankitaluthra1) [inline internal/storage/bucket_handle.go:486]: change level to trace

Comment (ankitaluthra1) [reply internal/storage/bucket_handle.go:486]: any error should be ignored here, please ad a comment on why its ignored

File: tools/integration_tests/util/setup/setup.go [hunk 2]
```diff
@@ -404,6 +405,27 @@ func AreBothMountedDirectoryAndTestBucketFlagsSet() bool {
...
+	}
+	if attrs.HierarchicalNamespace.Enabled {
+		mountConfig4 := config.MountConfig{
+			EnableHNS: true,
+			LogConfig: config.LogConfig{
+				Severity:        config.TRACE,
+				LogRotateConfig: config.DefaultLogRotateConfig(),
+			},
+		}
+		filePath4 := YAMLConfigFile(mountConfig4, "config_hns.yaml")
```
Comment (ankitaluthra1) [inline tools/integration_tests/util/setup/setup.go:422]: add a comment here "making config flag explicitly true when tests are run on hns bucket"

**Review by ankitaluthra1 — CHANGES_REQUESTED** [review]

File: internal/storage/bucket_handle.go (hunk 1 above)
Comment (Tulsishah) [reply internal/storage/bucket_handle.go]: Done

File: tools/integration_tests/util/setup/setup.go (hunk 2 above)
Comment (Tulsishah) [reply tools/integration_tests/util/setup/setup.go:423]: Done

File: internal/fs/inode/dir.go [hunk 3]
```diff
@@ -836,9 +836,11 @@ func (d *dirInode) DeleteChildDir(
 
 	// if the directory is an implicit directory, then no backing object
 	// exists in the gcs bucket, so returning from here.
-	if isImplicitDir {
```
Comment (ankitaluthra1) [inline internal/fs/inode/dir.go:839]: This method should be refactored and isImplicitDir check should be removed, delete method's behaviour should only be for delete, it should not have intelligence to know if passed directory is implicit or not, the caller passing isImplicitDir check is already aware that given directory is implicit or not, it should not call this method in case implicit dir is true.

File: internal/fs/inode/dir.go [hunk 4]
```diff
@@ -850,18 +852,18 @@ func (d *dirInode) DeleteChildDir(
 			Generation: 0, // Delete the latest version of object named after dir.
 		})
 
-	if err != nil {
-		err = fmt.Errorf("DeleteObject: %w", err)
-		return
-	}
-
 	if d.bucket.BucketType() == gcs.Hierarchical {
+		// In Hierarchical bucket implicit object remains as a folder only.
```
Comment (ankitaluthra1) [inline internal/fs/inode/dir.go:856]: Update this comment to specify why err is ignored in hns case

**Review by ankitaluthra1 — CHANGES_REQUESTED** [review]

File: internal/fs/inode/dir.go (hunk 3 above)
Comment (Tulsishah) [reply internal/fs/inode/dir.go:839]: As discussed ### offline, we considered implementing this logic outside the existing structure. However, we realized that this would necessitate a hierarchical bucket check due to the lack of implicit directories in such buckets. This check would need to be duplicated in various places, leading us to abandon this approach.

File: internal/fs/inode/dir.go (hunk 4 above)
Comment (Tulsishah) [reply internal/fs/inode/dir.go]: Done.

File: internal/gcsx/prefix_bucket_test.go [hunk 5]
```diff
@@ -418,3 +418,33 @@ func (t *PrefixBucketTest) GetFolder_Prefix() {
 	AssertEq("projects/_/buckets/some_bucket/folders/foo_something", result.GetName())
 
 }
+
+func (t *PrefixBucketTest) DeleteFolder() {
+	var err error
+	suffix := "taco"
+	name := t.prefix + suffix
+	contents := "foobar"
+
```
Comment (kislaykishore) [inline internal/gcsx/prefix_bucket_test.go:427]: Tests should have three visual blocks: Arrange, Act and Assert. They should be separated by space. Given that the object creation in the following lines is also part of the setup, we should have it in the same visual block. In other words, please remove this blank line.

File: internal/gcsx/prefix_bucket_test.go [hunk 6]
```diff
@@ -418,3 +418,33 @@ func (t *PrefixBucketTest) GetFolder_Prefix() {
...
+	// Create an object through the back door.
+	_, err = storageutil.CreateObject(t.ctx, t.wrapped, name, []byte(contents))
+	AssertEq(nil, err)
+
+	// Delete it.
+	err = t.bucket.DeleteFolder(
+		t.ctx,
+		suffix)
+
+	AssertEq(nil, err)
```
Comment (kislaykishore) [inline internal/gcsx/prefix_bucket_test.go:438]: The code henceforth is in the Assert zone. So let's remove all the blank lines here on through to the end of the test.

**Review by kislaykishore — CHANGES_REQUESTED** [review]
* Please explain the rationale behind the refactoring.
* The PR description talks about setting includeFolderAsPrefix as true. I'm not sure which method call accepts this value and is set to true. Could you explain that.

File: tools/integration_tests/util/setup/implicit_and_explicit_dir_setup/implicit_and_explicit_dir_setup.go [hunk 7]
```diff
@@ -64,7 +64,10 @@ func RunTestsForImplicitDirAndExplicitDir(flags [][]string, m *testing.M) int {
 }
 
 func RemoveAndCheckIfDirIsDeleted(dirPath string, dirName string, t *testing.T) {
-	operations.RemoveDir(dirPath)
+	err := os.RemoveAll(dirPath)
+	if err != nil {
+		t.Errorf("os.RemoveAll(): %v", err)
```
[the remark this answers is not in the document]
Comment (Tulsishah) [reply tools/integration_tests/util/setup/implicit_and_explicit_dir_setup/implicit_and_explicit_dir_setup.go:69]: This verb can be used only in 'fmt.Errorf' calls (%w)

File: internal/fs/inode/dir.go [hunk 8]
```diff
@@ -850,18 +852,19 @@ func (d *dirInode) DeleteChildDir(
...
 	if d.bucket.BucketType() == gcs.Hierarchical {
-		// Delete Folder deletes folder (in case of Hierarchical Bucket).
+		// Ignoring delete object error here, as in case hns there is no way of knowing
+		// if underlying placeholder object exists or not
+		// The DeleteFolder operation handles removing empty folders.
 		err = d.bucket.DeleteFolder(ctx, childName.GcsObjectName())
+		if err != nil {
+			return fmt.Errorf("DeleteFolder: %w", err)
+		}
+		time.Sleep(50 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (Tulsishah) [reply internal/fs/inode/dir.go]: Sorry.. I removed it but I think while rebasing something went wrong.

File: internal/gcsx/prefix_bucket_test.go (hunk 5 above)
Comment (Tulsishah) [reply internal/gcsx/prefix_bucket_test.go:427]: The other tests were in these format only. Although I fixed mine.

File: internal/config/yaml_parser.go [hunk 9]
```diff
@@ -186,5 +186,10 @@ func ParseConfigFile(fileName string) (mountConfig *MountConfig, err error) {
 		return mountConfig, fmt.Errorf("error parsing list config: %w", err)
 	}
 
+	// In this scenario, the value for HNS should be overridden to true.
```
[the remark this answers is not in the document]
Comment (Tulsishah) [reply internal/config/yaml_parser.go]: Update the comment.

File: internal/fs/inode/dir.go [hunk 10]
```diff
@@ -850,18 +852,19 @@ func (d *dirInode) DeleteChildDir(
...
-	}
-
 	if d.bucket.BucketType() == gcs.Hierarchical {
-		// Delete Folder deletes folder (in case of Hierarchical Bucket).
+		// Ignoring delete object error here, as in case hns there is no way of knowing
+		// if underlying placeholder object exists or not
+		// The DeleteFolder operation handles removing empty folders.
 		err = d.bucket.DeleteFolder(ctx, childName.GcsObjectName())
+		if err != nil {
+			return fmt.Errorf("DeleteFolder: %w", err)
```
[the remark this answers is not in the document]
Comment (Tulsishah) [reply internal/fs/inode/dir.go:860]: We (me and @ankitaluthra1) discussed this and reached to conclusion for  Hierarchical bucket We concluded that we can safely ignore errors from DeleteObject since we also call DeleteFolder, which will return any err if occur.

File: internal/gcsx/prefix_bucket_test.go [hunk 11]
```diff
@@ -418,3 +418,33 @@ func (t *PrefixBucketTest) GetFolder_Prefix() {
 	AssertEq("projects/_/buckets/some_bucket/folders/foo_something", result.GetName())
 
 }
+
+func (t *PrefixBucketTest) DeleteFolder() {
```
[the remark this answers is not in the document]
Comment (Tulsishah) [reply internal/gcsx/prefix_bucket_test.go]: Done

**Comment by Tulsishah** [comment]
> * Please explain the rationale behind the refactoring.
> * The PR description talks about setting includeFolderAsPrefix as true. I'm not sure which method call accepts this value and is set to true. Could you explain that.

Added

File: tools/integration_tests/util/setup/implicit_and_explicit_dir_setup/implicit_and_explicit_dir_setup.go [hunk 12]
```diff
@@ -64,7 +64,10 @@ func RunTestsForImplicitDirAndExplicitDir(flags [][]string, m *testing.M) int {
...
+	}
```
Comment (kislaykishore) [inline tools/integration_tests/util/setup/implicit_and_explicit_dir_setup/implicit_and_explicit_dir_setup.go:70]:
```suggestion
	require.Nil(err)
```

File: tools/integration_tests/util/setup/setup.go [hunk 13]
```diff
@@ -404,6 +405,29 @@ func AreBothMountedDirectoryAndTestBucketFlagsSet() bool {
 	return false
 }
 
+// Explicitly set the enable-hns config flag to true when running tests on the HNS bucket.
+func AddHNSFlagForHierarchicalBucket(ctx context.Context, storageClient *storage.Client) (flags []string) {
```
Comment (kislaykishore) [inline tools/integration_tests/util/setup/setup.go:409]: Fix naked return. Return error from the function.

File: tools/integration_tests/operations/operations_test.go [hunk 14]
```diff
@@ -189,6 +176,14 @@ func TestMain(m *testing.M) {
 		flagsSet = append(flagsSet, []string{"--client-protocol=grpc", "--implicit-dirs=true"})
 	}
 
+	// HNS tests utilize the gRPC protocol, which is not supported by TPC.
+	if !setup.TestOnTPCEndPoint() {
+		hnsFlagConfig := setup.AddHNSFlagForHierarchicalBucket(ctx, storageClient)
+		if hnsFlagConfig != nil {
+			flagsSet = append(flagsSet, hnsFlagConfig)
+		}
```
Comment (kislaykishore) [inline tools/integration_tests/operations/operations_test.go:184]:
```suggestion
		if hnsFlagConfig := setup.AddHNSFlagForHierarchicalBucket(ctx, storageClient); hnsFlagConfig != nil {
			flagsSet = append(flagsSet, hnsFlagConfig)
		}
```

**Review by kislaykishore — CHANGES_REQUESTED** [review]

**Review by kislaykishore — APPROVED** [review]
Please get it reviewed by someone else on the team as well.

**Review by ankitaluthra1 — APPROVED** [review]

File: tools/integration_tests/util/setup/setup.go [hunk 15]
```diff
@@ -404,6 +405,30 @@ func AreBothMountedDirectoryAndTestBucketFlagsSet() bool {
...
+	}
+	if !attrs.HierarchicalNamespace.Enabled {
+		return nil, fmt.Errorf("Bucket is not Hierarchical")
+	}
+
+	var flags []string
+	mountConfig4 := config.MountConfig{
+		EnableHNS: true,
+		LogConfig: config.LogConfig{
+			Severity:        config.TRACE,
```
Comment (charith87) [inline tools/integration_tests/util/setup/setup.go:422]: Why is this needed?

**Review by charith87 — APPROVED** [review]

File: tools/integration_tests/util/setup/setup.go (hunk 15 above)
Comment (Tulsishah) [reply tools/integration_tests/util/setup/setup.go:422]: It is for integration tests. So in case of failure we can see logs.
